its/SonarCloudIntegrationTests.cs

```diff
@@ -0,0 +1,192 @@
+        }
+
+        private static VssConnection GetAzDoConnection()
+        {
+            VssBasicCredential credentials = new VssBasicCredential(string.Empty, Environment.GetEnvironmentVariable("AZURE_TOKEN"));
+            VssConnection connection = new VssConnection(new Uri(String.Concat(Environment.GetEnvironmentVariable("AZDO_BASE_URL"), Environment.GetEnvironmentVariable("AZDO_ITS_ORGA"))), credentials);
+
+            return connection;
+        }
+
+        private async Task AssertNcLocAndCoverage(String projectKey, long expectedNcLoc, double expectedCoverage)
+        {
+            //Checking results on SonarQube's side. As all pipelines have a Publish Quality Gate Result in them, 
+            //we don't need to check if background task is finished, because it should be by then.
+            var actualNcLoc = await _scInstance.GetNclocForProjectAsync(projectKey);
+            var actualCoverage = await _scInstance.GetCodeCoveragePercentageForProjectAsync(projectKey);
+
+            Assert.AreEqual(expectedNcLoc, actualNcLoc);
+            Assert.AreEqual(expectedCoverage, actualCoverage);
+        }
+
+        private async Task AssertVersionsOfTasksAreCorrect(Build currentBuildResult, bool shouldAssertAnalyze = true)
+        {
+            var buildProperties = await _buildHttpClient.GetBuildPropertiesAsync(Environment.GetEnvironmentVariable("AZDO_ITS_PROJECT_NAME"), currentBuildResult.Id);
+            var actualPrepareVersion = buildProperties.FirstOrDefault(p => p.Key.Equals("SonarSourcePrepareTaskVersion"));
+            var actualAnalyzeVersion = buildProperties.FirstOrDefault(p => p.Key.Equals("SonarSourceAnalyzeTaskVersion"));
+            var actualPublishVersion = buildProperties.FirstOrDefault(p => p.Key.Equals("SonarSourcePublishTaskVersion"));
+
+            var expectedPrepareVersion = System.Environment.GetEnvironmentVariable("PREPARE_TASK_VERSION");
+            var expectedAnalyzeVersion = System.Environment.GetEnvironmentVariable("ANALYZE_TASK_VERSION");
+            var expectedPublishVersion = System.Environment.GetEnvironmentVariable("PUBLISH_TASK_VERSION");
+
+            Assert.AreEqual(expectedPrepareVersion, actualPrepareVersion.Value.ToString());
+
+            if (shouldAssertAnalyze)
+            {
+                Assert.AreEqual(expectedAnalyzeVersion, actualAnalyzeVersion.Value.ToString());
+            }
+
+            Assert.AreEqual(expectedPublishVersion, actualPublishVersion.Value.ToString());
+                
+        }
+
+        private async Task<Build> ExecuteBuildAndWaitForCompleted(string pipelineName)
+        {
+            var definitions = await _buildHttpClient.GetDefinitionsAsync(project: Environment.GetEnvironmentVariable("AZDO_ITS_PROJECT_NAME"));
+            var target = definitions.FirstOrDefault(d => d.Name == pipelineName);
+
+            var queuedBuild = await _buildHttpClient.QueueBuildAsync(new Build
```

Comment:
where are the projects that are being built?

Reply:
sample repository in sonarsource-integration-tests organization on AzDo

Reply:
Could you document the overall structure of the test environment please? i.e. what the various organisations and pipelines are used for. It's hard to understand the big picture just by looking at the yaml/code.

Reply:
Here's the doc : https://xtranet-sonarsource.atlassian.net/wiki/spaces/SSG/pages/2298675258/Integration+Tests